### PR TITLE
React/dp 9399  add backstop

### PR DESCRIPTION
**Pull request by pmallett-mc**
Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/dev/docs/change-log-instructions.md)

## Description
Adds backstop tests for Mayflower/react

## Related Issue / Ticket
https://jira.mass.gov/browse/DP-9399

## Steps to Test
* `git checkout react/DP-9399--add-backstop`
* `nvm use && npm install`
* `npm run build-storybook` # the backstop tests will access the static build of the Storybook
* `npm run backstop`

* Also tests in Circle on this PR should all pass

## Screenshots
n/a

## Additional Notes:
n/a

#### Impacted Areas in Application
* added npm run commands for backstop testing in /react

#### @TODO
* add CHANGELOG

### CODE REVIEW

File: .circleci/config.yml
Comment (rbayliss) [inline .circleci/config.yml:123]: This will fail if either of these tests fails, but splitting into two commands will cause the second report to be the one that's saved to the artifacts and reported on as JUnit.   I think it would probably be better to have an `--all` flag (or make that the default) to run them all in a single run so the report has everything.

Comment (rbayliss) [reply .circleci/config.yml:123]: This goes for `npm run backstop` as well - the report will only reflect the most recent one run.  I'd suggest finding a way to make the flags combine so `--atoms --other` will run all of the tests.   Are you sure you don't want to just combine these into a single test?  It seems like it'll make your workflow simpler - yes, it's a bit slower, but it seems worth it.

File: react/docs/visual-regression.md
Comment (rbayliss) [inline react/docs/visual-regression.md]: I think this section needs updating - backstop.atoms.js doesn't exist anymore.

File: react/backstop/backstop.js
Comment (rbayliss) [inline react/backstop/backstop.js:92]: Just wanted to note that we've seen some issues hitting resource limits with these async comparisons.  I think in our case it has a lot to do with the pixel sizes of the pages we're capturing and comparing, so I wouldn't change them now, but keep in mind that the limits may need to drop if you start getting reports of hanging or oddly failing tests.

**Review by rbayliss — APPROVED** [review]
This is looking really good!  The one thing I'd suggest addressing is the reporting (explained below).  Other than that, I think you're in good shape here!

**Comment by mrjmd** [comment]
Talked to @petermallett, this is still a WIP so holding off doing a full review until the work is complete, removing label for now.

File: react/backstop/backstop.js
Comment (pmallett-mc) [reply react/backstop/backstop.js:92]: Yea, I noticed they had to be lowered quite a bit for the PL build; I think our viewport sizes are small enough that this will run fine for now, though.

File: react/docs/visual-regression.md
Comment (pmallett-mc) [reply react/docs/visual-regression.md]: Updated docs/visual-regression.md in 475c7c3f4

File: .circleci/config.yml
Comment (pmallett-mc) [reply .circleci/config.yml:123]: I agree simplifying this is probably desired & I will test the speed differences in the next round of updates for this. For now I think we're going to add them as-is so we can unblock work on sharing SCSS a bit.

I set it up in a single run step so that if --atoms fails, that is the report that will be written, the second test config will not be run.

Again, agreed the complexity is probably not worth the speed savings overall & will probably be cleaned up in the next update.

File: react/.eslintignore
Comment (mrjmd) [inline react/.eslintignore:6]: We eslintignore tests on search as well, but I'm not sure if it's something we should be considering changing long term. It definitely had led to things slipping through the cracks (unused variables & imports, hanging spaces, etc). Thoughts?

File: react/backstop/backstop.js
Comment (mrjmd) [inline react/backstop/backstop.js:26]: How arbitrary is the choice of viewport size here, or what motivated it?

File: react/backstop/scripts/onBefore.js
Comment (mrjmd) [inline react/backstop/scripts/onBefore.js]: I'm assuming this file is left here for future use?

**Review by mrjmd — APPROVED** [review]
All tests pass, however the readme instructions are a little unclear / didn't quite work for me as written. I tried to just run `npm run backstop` since it shows `npm-run-all` as the first step, however this didn't work. I then tried running `npm run build-storybook` as the prerequisite mentioned in the readme, however this failed because css assets weren't built. So to run tests my actual process was:

- cd react
- npm i
- npm run build-css
- npm run build-storybook
- npm run backstop

After that, everything worked fine.

**Comment by pmallett-mc** [comment]
Temporarily re-adding the WIP label so we can test combined PL/react build/test

File: react/backstop/scripts/onBefore.js
Comment (pmallett-mc) [reply react/backstop/scripts/onBefore.js]: Yea, though I can't imagine we will ever need to add any cookies; maybe should just remove it.

**Review by clairesunstudio — APPROVED** [review]
works great!

File: react/.eslintignore
Comment (pmallett-mc) [reply react/.eslintignore:6]: I agree we should probably lint tests, too -- I'll note in the ticket for followup.

Comment (pmallett-mc) [reply react/.eslintignore:6]: Also note: this is for the test that came with (I think?) nwb, not the backstop code in react/backstop

File: react/backstop/backstop.js
Comment (pmallett-mc) [reply react/backstop/backstop.js:26]: It's large enough to fit all the atoms; See DP-9399 todo's for followup.

**Pull request merged**
